apiserver/remoterelations/remoterelations.go

```diff
@@ -184,55 +185,136 @@ func (api *RemoteRelationsAPI) RelationUnitSettings(relationUnits params.Relatio
-		tag, err := names.ParseRelationTag(entity.Tag)
+	one := func(relTag params.Entity) (*params.IngressSubnetInfo, error) {
+		// Load the relation details for the current tag.
+		remoteRelation, err := api.remoteRelation(relTag)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		rel, err := api.st.KeyRelation(tag.Id())
+
+		// Double check we have a source model - we should never hit this.
+		if remoteRelation.SourceModelUUID == "" {
+			return nil, errors.NotValidf("remote relation %v with empty source model UUID", remoteRelation)
+		}
+
+		var result params.IngressSubnetInfo
+		result.ApplicationName = remoteRelation.ApplicationName
+
+		// Until networking support becomes more sophisticated, for now
+		// we only care about opening access to applications with an endpoint
+		// having the "provider" role. The assumption is that such endpoints listen
+		// to incoming connections and thus require ingress. An exception to this
+		// would be applications which accept connections onto an endpoint which
+		// has a "requirer" role.
+		if remoteRelation.Endpoint.Role != charm.RoleProvider {
+			logger.Debugf(
+				"no ingress network for application %v without provides endpoint %v",
+				remoteRelation.ApplicationName, remoteRelation.Endpoint.Name)
+			return &result, nil
+		}
+
+		// Get the source model of the remote application so we can determine the
+		// required CIDRs for which we need to allow ingress to this model.
+		sourceSt, closer, err := api.pool.Get(remoteRelation.SourceModelUUID)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		result := &params.RemoteRelation{
-			Id:   rel.Id(),
-			Life: params.Life(rel.Life().String()),
-			Key:  tag.Id(),
+		defer closer()
+
+		// TODO(wallyworld) - as a first implementation, just get all CIDRs from the model
+		all, err := sourceSt.AllSubnets()
+		cidrs := set.NewStrings()
+		if err != nil {
+			return nil, errors.Trace(err)
 		}
-		for _, ep := range rel.Endpoints() {
-			// Try looking up the info for the remote application.
-			_, err := api.st.RemoteApplication(ep.ApplicationName)
-			if err != nil && !errors.IsNotFound(err) {
+
+		for _, subnet := range all {
+			ip, _, err := net.ParseCIDR(subnet.CIDR())
+			if err != nil {
 				return nil, errors.Trace(err)
-			} else if err == nil {
-				result.RemoteEndpointName = ep.Name
+			}
+			if ip.IsLoopback() || ip.IsMulticast() {
 				continue
 			}
-			// Try looking up the info for the local application.
-			_, err = api.st.Application(ep.ApplicationName)
-			if err != nil && !errors.IsNotFound(err) {
-				return nil, errors.Trace(err)
-			} else if err == nil {
-				result.ApplicationName = ep.ApplicationName
-				result.Endpoint = params.RemoteEndpoint{
-					Name:      ep.Name,
-					Interface: ep.Interface,
-					Role:      ep.Role,
-					Scope:     ep.Scope,
-					Limit:     ep.Limit,
-				}
+			// TODO(walluworld) - We only support IPv4 addresses as not all providers support IPv6.
+			if ip.To4() == nil {
```

Comment:
I'll fix this properly in a follow up, once the firewaller worker has gained a little more capability